Add support for arm64-linux (#1)

achichinov opened the pull request:
Add support for arm64-linux

- Copied the latest config.guess
- Copied the latest linux_syscall_support.h from google LSS repository with __aarch64__ support
  Merged some changes from existing coredumper linux_syscall_support.h that are required but absent in the latest LSS.
- Added preprocessor conditions for __aarch64__
- Copied the GP and FP registers structures from glibc.
- Defined FRAME and SET_FRAME macros
  FRAME is defined via getcontext() function. It is simpler than using asm.
- In CreateElfCore() skip [vvar] memory segment.
  A process fails to read this segment. On x86_64 [vvar] is disabled for core dump, but on ARM64 it is not.
  It is the only segment that is marked as "pure PFN range" on the both architectures.
- Make stack pointer 16-byte aligned in local_clone() on all architectures.
- Add a separate ifdef section to read registers using PTRACE_GETREGSET.
  PTRACE_GETREGS and PTRACE_GETFPREGS are not supported on arm64-linux

achichinov committed:
Add support for arm64-linux (#1)

Add support for arm64-linux

- Added preprocessor conditions for aarch64
- Copied the GP and FP registers structures from glibc.
- Defined FRAME and SET_FRAME macros
  FRAME is defined via getcontext() function. It is simpler than using asm.
- In CreateElfCore() skip [vvar] memory segment.
  A process fails to read this segment. On x86_64 [vvar] is disabled for core dump, but on ARM64 it is not.
  It is the only segment that is marked as "pure PFN range" on the both architectures.
- Make stack pointer 16-byte aligned in local_clone() on all architectures.
- Add a separate ifdef section to read registers using PTRACE_GETREGSET.
  PTRACE_GETREGS and PTRACE_GETFPREGS are not supported on arm64-linux

Commit:
Check for pf flag only on arm64.

madscientist commented on config.guess:
```diff
@@ -42,7 +42,7 @@ timestamp='2021-06-03'
 
 
-me=`echo "$0" | sed -e 's,.*/,,'`
+me=$(echo "$0" | sed -e 's,.*/,,')
```
I'm assuming that these changes were made when you used some different version of autoconf, and were not made by hand by you?

achichinov replied:
> I'm assuming that these changes were made when you used some different version of autoconf, and were not made by hand by you?

I copied the whole file, but it looks like I copied a modified version. And I don't recall where I copied it from. Most probably from github.
I just checked this file on GNU repo and it does not have these modifciations
https://git.savannah.gnu.org/gitweb/?p=config.git;a=blob_plain;f=config.guess;hb=45e181800a6a27268a9c5d79dcc60492fef9a9a0

I will update the file from the official repo.

madscientist commented on src/elfcore.c:
```diff
@@ -73,6 +73,8 @@ extern "C" {
   #elif defined(__ARM_ARCH_3__)
     #define O_LARGEFILE 0400000
+  #elif defined(__aarch64__)
+    #define O_LARGEFILE  __O_LARGEFILE
```
I wonder if it's worthwhile to use something like, **#elif defined (__O_LARGEFILE) \n #define O_LARGEFILE __O_LARGEFILE** just to get all platforms that do define __O_LARGEFILE.

This works too.

achichinov replied:
I think this is good idea. I will do it.

madscientist commented on src/elfcore.c:
```diff
@@ -157,13 +168,15 @@ typedef struct elf_timeval {    /* Time value with microsecond resolution    */
-
+#else
+typedef struct elf_siginfo elf_siginfo;
+#endif
```
Strange.  Why isn't this defined on Intel?

achichinov replied:
> Strange. Why isn't this defined on Intel?

It is defined for Intel too in `<sys/procfs.h>` which was not included directly. But it was picked up on arm64 somehow.
I will include `<sys/procfs.h>` and remove this `#if`

madscientist commented on src/elfcore.c:
```diff
@@ -1750,6 +1779,28 @@ int InternalGetCoreDump(void *frame, int num_threads, pid_t *pids,
+      }
+      memset(scratch, 0xFF, sizeof(scratch));
+      scratch_iovec.iov_len = sizeof(scratch);
+      if (sys_ptrace(PTRACE_GETREGSET, pids[i], (void*)NT_FPREGSET, &scratch_iovec) == 0) {
```
ptrace() defines the return code to be -1 on error.  Is it better to check for >=0 here?  Or invert the test and check for <0 and use the goto in the then-block, then you don't need the else statement at all.

madscientist commented on src/elfcore.c:
```diff
@@ -1785,10 +1836,12 @@ int InternalGetCoreDump(void *frame, int num_threads, pid_t *pids,
   /* Get parent's CPU registers, and user data structure                     */
   {
     #ifndef __mips__
+    #ifndef __aarch64__ // PTRACE_PEEKUSER returns -1 and set errno = 5 (EIO 5 Input/output error)
```
What are the ramifications of not setting this **user** value?

Also, did you see this note from the ptrace(2) man page:

>    C library/kernel differences
       At the system call level,  the  PTRACE_PEEKTEXT,  PTRACE_PEEKDATA,  and
       PTRACE_PEEKUSER requests have a different API: they store the result at
       the address specified by the data parameter, and the  return  value  is
       the  error  flag.  The glibc wrapper function provides the API given in
       DESCRIPTION above, with the result  being  returned  via  the  function
       return value.

So, maybe it didn't fail after all?

achichinov replied:
I just realized that this is  `sys_ptrace` from `linux_syscall_support.h`. I tested the results on both platforms.
It returns data on x86 for some offsets and it always returns -1 (EIO) on arm64.
However, I will remove this `#ifdef` and revert the change back to the original code to avoid unforeseen results.

madscientist commented on src/elfcore.c:
```diff
@@ -1806,7 +1859,9 @@ int InternalGetCoreDump(void *frame, int num_threads, pid_t *pids,
   prpsinfo.pr_gid   = sys_getegid();
   prpsinfo.pr_pid   = main_pid;
   prpsinfo.pr_ppid  = sys_getppid();
+  #if !defined(__aarch64__)
```
Strange, why wouldn't getpgrp() not be available on ARM?  That's a bog-standard POSIX call that's been around forever.

achichinov replied:
> Strange, why wouldn't getpgrp() not be available on ARM? That's a bog-standard POSIX call that's been around forever.

This is a syscall `sys_getpgrp()` from `LSS`. It is deprecated in the latest `asm-generic/unistd.h` but can be enabled via `__ARCH_WANT_SYSCALL_DEPRECATED`
I will enable it.

achichinov replied:
It turned out that `getpgrp()` always returns the PGID, so don't need to bother with the system calls from LSS and use `getpgrp()` on all platforms.

madscientist commented on src/elfcore.c:
```diff
@@ -1897,8 +1952,8 @@ int InternalGetCoreDump(void *frame, int num_threads, pid_t *pids,
-    int openmax  = sys_sysconf(_SC_OPEN_MAX);
-    int pagesize = sys_sysconf(_SC_PAGESIZE);
+    int openmax  = sysconf(_SC_OPEN_MAX);
+    int pagesize = sysconf(_SC_PAGESIZE);
```
Why this change?

achichinov replied:
The latest `linux_syscall_support.h` does not provide `sys_sysconf()`

madscientist commented on src/linux_syscall_support_unittest.cc:
```diff
@@ -140,42 +140,73 @@ static void SigAction(int signum, siginfo_t *si, void *arg) {
+    CHECK(!sys_rt_sigaction(signum, NULL, &orig, kSigsetSize));
+    #else
     CHECK(!sys_sigaction(signum, NULL, &orig));
+    #endif
```
Is there a reason we can't just switch all the architectures to use rt_sigaction()?  It should work on Intel shouldn't it?

achichinov replied:
`linux_syscall_support_unittest` fails with SIGSEGV when calling `sys_rt_sigsuspend()` on Intel.
And it looks like `rt` functions can not be mixed with non `rt`.

madscientist commented on src/linuxthreads.c:
```diff
@@ -609,16 +618,22 @@ int ListAllProcessThreads(void *parameter,
+      while ((rc = sys_waitpid(clone_pid, &status, __WALL)) < 0 && ERRNO == EINTR) {
+             /* Keep waiting                                                 */
+      }
+      #else
```
Why does this need to be different?

achichinov replied:
One is `sys_waitpid()` and another is `sys0_waitpid()`. I don't recall why I added It.
Original code works on the both platforms, so I will revert the change.

madscientist commented on src/linux_syscall_support.h:
```diff
@@ -1,4 +1,4 @@
-/* Copyright (c) 2005-2008, Google Inc.
+/* Copyright (c) 2005-2011, Google Inc.
```
I didn't review this, I hope it's right :)

achichinov replied:
The version at 2011 is so perfect that no new versions are necessary. :-)

madscientist reviewed (approved):
Looks good although it seems like there are some simplifications that could be made to reduce the amount of ifdefs

achichinov committed:
Fixes for PR (#2)

* Replace config.guess with the one from the official GNU repo
* use __O_LARGEFILE if it si defined. use struct elf_siginfo from procfs.h on all archs
* run PTRACE_PEEKUSER on arm64
* simplified if statements in the reading of the registers for arm64.
* use getpgrp() instead of sys_getpgrp() which is deprecated on arm64
* sys0_waitpid() works fine on arm64-linux

madscientist commented on src/elfcore.c:
```diff
@@ -923,6 +934,17 @@ static int CreateElfCore(void *handle,
                     }
 
+                    #if defined(__aarch64__)
+                    /* check if the flag is "pf" */
```
This comment is not helpful since it's obvious that this is what the code does.  Better would be to say WHY we check for this; apparently we don't dump segments with the "pf" flag for some reason?

madscientist commented on src/elfcore.c:
```diff
@@ -1806,7 +1847,7 @@ int InternalGetCoreDump(void *frame, int num_threads, pid_t *pids,
   prpsinfo.pr_pid   = main_pid;
   prpsinfo.pr_ppid  = sys_getppid();
-  prpsinfo.pr_pgrp  = sys_getpgrp();
+  prpsinfo.pr_pgrp  = getpgrp();
```
Just curious why we are using this rather than the sys_ version?  Maybe a comment here explaining will help me and future readers.

madscientist commented on src/elfcore.c:
```diff
@@ -1897,8 +1938,8 @@ int InternalGetCoreDump(void *frame, int num_threads, pid_t *pids,
-    int openmax  = sys_sysconf(_SC_OPEN_MAX);
-    int pagesize = sys_sysconf(_SC_PAGESIZE);
+    int openmax  = sysconf(_SC_OPEN_MAX);
+    int pagesize = sysconf(_SC_PAGESIZE);
```
Ditto here: why do we remove sys?

madscientist commented on src/linuxthreads.c:
```diff
@@ -316,9 +321,13 @@ static void ListerThread(struct ListerParams *args) {
+    sys_rt_sigaction(sync_signals[sig], &sa, (struct kernel_sigaction *)NULL, kSigsetSize);
+    #else
     sys_sigaction(sync_signals[sig], &sa, (struct kernel_sigaction *)NULL);
+    #endif
```
Is there a reason we can't use rt_sigaction() on intel (that is, just change this unconditionally to use rt_sigaction?)

it's OK if so, I just wonder why.

achichinov committed:
More updates (#3)

 - expands a comment about pf flag
 - adds comments about replacing sys_getpgrp and sys_sysconf with getpgrp and sysconf
 - calls sys_rt_sigaction on both platforms in linuxthreds.c

madscientist reviewed (approved):
Looks good, thanks.  Are you ready for it to be pulled?

achichinov commented:
> Looks good, thanks. Are you ready for it to be pulled?

Yes.
